src/library/scala/collection/mutable/ArrayDeque.scala

```diff
@@ -477,24 +476,33 @@ class ArrayDeque[A] protected (
   /**
     * Trims the capacity of this ArrayDeque's instance to be the current size
     */
-  def trimToSize(): Unit = resize(length - 1)
+  def trimToSize(): Unit = resize(length)
+
+  private[mutable] def capacityForTesting: Int = array.length
 
   // Utils for common modular arithmetic:
   @inline protected def start_+(idx: Int) = (start + idx) & (array.length - 1)
   @inline private[this] def start_-(idx: Int) = (start - idx) & (array.length - 1)
   @inline private[this] def end_+(idx: Int) = (end + idx) & (array.length - 1)
   @inline private[this] def end_-(idx: Int) = (end - idx) & (array.length - 1)
 
-  @inline private[this] def isResizeNecessary(len: Int) = {
-    // Either resize if we need more cells OR we need to downsize BUT not a good idea to repeatedly resize small arrays
-    len >= array.length || (2*len < array.length && array.length >= ArrayDeque.StableSize)
+  @inline private[this] def mustGrow(len: Int) = {
+    len >= array.length
   }
 
+  @inline private[this] def shouldShrink(len: Int) = {
+    // To avoid allocation churn, only shrink when array is large
+    // and less than 2/5 filled.
```

Comment:
why 2/5 not 1/2? Is this a common practice?

Reply:
You want to have your contraction threshold smaller than your expansion threshold because you want to avoid the case where you have, say, 2^m elements and add one to get 2^m + 1, causing an allocation of 2^(m+1) slots; then you remove one to go back to 2^m, and you reallocate 2^m slots.

Exactly what the threshold is varies, with no clear theory (or pragmatic data that I know of) on what is "best".  (There is a similar conundrum for fill fractions for open hashing schemes.)  One common one is 1/4, which gives a worst-case scenario of 2^(m-1) + 2^m + 2^(m+1) allocations for a cycle that removes and then adds back 2^(m-1)+1 elements (starting from 2^m+1).  But a worst case scenario for wasted space is that you consume 4x the memory that is strictly required.

In contrast, a 2/5 rule gives a worst-case reallocation of 2^m + 2^(m+1) elements after a cycle of length 2^(m-1)/5.  However, at worst you can consume 2.5x the memory that you actually need.

Is the allocation ratio of ~7x the cycle length worth the 4x memory danger?  Or is the potential 30x allocation ratio of the 2/5 scheme implausible enough so it's worth capping memory only 2.5x?  Kinda hard to say.

I don't have any particular advice, despite all the typing.

Reply:
Happy to change this further if you’d suggest...

Reply:
Well you certainly need to fix the numeric overflow in `((5 * len) >>> 1)`.  `len` might be 2^30-2.  Dividing by 5 is kinda expensive, though.  (Also, you always realloc on 0; is this intentional?  It probably shouldn't be.)

Reply:
I think `array.length  - 5 * (len >>> 1)  > 0` fixes the overflow in most cases?

> you always realloc on 0

I'm not sure what you mean by this

Reply:
No, that still overflows in a bad way.  If array.length is 2^30 and len is 2^30-m where m is small, 5*(len >> 1) is roughly 2^31 + 2^29, which is -2^29 signed, and 2^30 - (-2^29) doesn't overflow and is bigger than 0.  But of course you shouldn't reallocate when removing just a few elements.  The overflow _actually_ needs to be fixed, not just "in most cases"!

I also mean that the condition is always met if `len` is 0.  But reallocating a 1-length array for a 0-length array seems a bad idea, even though 0 is, to be fair, smaller than 2/5.

Reply:
happy to fix the overflow issue, but to be clear I don't think my PR is introducing this bug.  The old logic checks for `2*len < array.length`.  

If `array.length` is 2^31 and len is `2^31 - m`, then `2 * len` will overflow and the check will pass.

Reply:
I thought the size of the collection was capped at 2^30 because of the bit shift mask.  You can't actually allocate an array of 2^31 (because in an `Int` 0x80000000 is -2^31), and you have to allocate powers of two.  So the old scheme avoided overflow, trickily, but probably on purpose.